crypto/core_namemap.c

```diff
@@ -216,62 +179,97 @@ int ossl_namemap_name2num_n(const OSSL_NAMEMAP *namemap,
     return ret;
 }
 
-struct num2name_data_st {
-    size_t idx;                  /* Countdown */
-    const char *name;            /* Result */
-};
-
-static void do_num2name(const char *name, void *vdata)
+const char *ossl_namemap_num2name(const OSSL_NAMEMAP *namemap, int number,
+                                  size_t idx)
 {
-    struct num2name_data_st *data = vdata;
+    NAMES *names;
+    const char *ret = NULL;
+
+    if (namemap == NULL || number <= 0)
+        return NULL;
 
-    if (data->idx > 0)
-        data->idx--;
-    else if (data->name == NULL)
-        data->name = name;
+    if (!CRYPTO_THREAD_read_lock(namemap->lock))
+        return NULL;
+
+    names = sk_NAMES_value(namemap->numnames, number - 1);
+    if (names != NULL)
+        ret = sk_STRING_value(names, idx);
+
+    CRYPTO_THREAD_unlock(namemap->lock);
+
+    return ret;
 }
 
-const char *ossl_namemap_num2name(const OSSL_NAMEMAP *namemap, int number,
-                                  size_t idx)
+/* This function is not thread safe, the namemap must be locked */
+static int numname_insert(OSSL_NAMEMAP *namemap, int number,
+                          const char *name)
 {
-    struct num2name_data_st data;
+    NAMES *names;
+    char *tmpname;
+
+    if (number > 0) {
+        names = sk_NAMES_value(namemap->numnames, number - 1);
+        if (!ossl_assert(names != NULL)) {
+            /* cannot happen */
+            return 0;
+        }
+    } else {
+        /* a completely new entry */
+        names = sk_STRING_new_null();
+        if (names == NULL)
+            return 0;
+    }
 
-    data.idx = idx;
-    data.name = NULL;
-    if (!ossl_namemap_doall_names(namemap, number, do_num2name, &data))
-        return NULL;
-    return data.name;
+    if ((tmpname = OPENSSL_strdup(name)) == NULL)
+        goto err;
+
+    if (!sk_STRING_push(names, tmpname))
+        goto err;
+
+    if (number <= 0) {
+        if (!sk_NAMES_push(namemap->numnames, names))
+            goto err;
+        number = sk_NAMES_num(namemap->numnames);
+    }
+    return number;
+
+ err:
+    if (number <= 0)
+        sk_STRING_free(names);
+    OPENSSL_free(tmpname);
+    return 0;
 }
 
 /* This function is not thread safe, the namemap must be locked */
 static int namemap_add_name(OSSL_NAMEMAP *namemap, int number,
                             const char *name)
 {
-    NAMENUM_ENTRY *namenum = NULL;
-    int tmp_number;
+    int ret;
+    HT_VALUE val = { 0 };
+    NAMENUM_KEY key;
 
     /* If it already exists, we don't add it */
-    if ((tmp_number = namemap_name2num(namemap, name)) != 0)
-        return tmp_number;
+    if ((ret = ossl_namemap_name2num(namemap, name)) != 0)
+        return ret;
 
-    if ((namenum = OPENSSL_zalloc(sizeof(*namenum))) == NULL)
+    if ((number = numname_insert(namemap, number, name)) == 0)
```

Comment:
I think this should be moved down below the call to ossl_ht_insert, so that we only push a name onto the stack if its successfully inserted into the hashtable, assuming the above call to name2num is removed)

Reply:
The numname_insert call is done first to return the _number_ to be inserted into the hash table. As we are under lock we could check what is the highest number in the numname mapping first, then try to ht_insert and only then do numname_insert. It has the disadvantage that we would have to remove the hash table entry if numname_insert for some reason fails (most probably an allocation error). However we do not lock the hash table on read so we actually cannot remove things from it. So no, this has to be kept as is.